### ISSUE TITLE
Create the obj-c extractor. [phab:D1038]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1377]
_Created by **salguarnieri** at 2016-09-06 21:40:36_: 

This is a small diff on the cxx extractor.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-06 21:45:56_:

  - make scripts executable

**Comment by kythe** [comment]
_**zarko** wrote at 2016-09-06 23:34:30_:

Is it possible to include a unit test that runs the extractor on a hello, worldish example?

Do we need to put the Bazel-owned scripts in `third_party`?

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-06 23:41:35_:

>>! In (earlier comment), zarko wrote:
> Is it possible to include a unit test that runs the extractor on a hello, worldish example?

I'll see if I can add one. I didn't see one for the cxx bazel extractor, should I create one for that also or did I miss it?

> Do we need to put the Bazel-owned scripts in `third_party`?
Maybe. I carved apart a single script from them (https://github.com/bazelbuild/bazel/blob/master/src/tools/xcode/xcrunwrapper/xcrunwrapper.sh) so I'm really unsure what is the right thing to do.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-09-07 00:00:33_:

>>! In (earlier comment), salguarnieri wrote:
>>>! In (earlier comment), zarko wrote:
>> Is it possible to include a unit test that runs the extractor on a hello, worldish example?
> 
> I'll see if I can add one. I didn't see one for the cxx bazel extractor, should I create one for that also or did I miss it?

You're right; it doesn't look like we have that hooked up. (I'd initially thought that `cc_verifier_test` uses it, but reality differs.) It would be better to keep this diff ObjC-only and to consider adding one of the cxx extractor separately.

>> Do we need to put the Bazel-owned scripts in `third_party`?
> Maybe. I carved apart a single script from them (https://github.com/bazelbuild/bazel/blob/master/src/tools/xcode/xcrunwrapper/xcrunwrapper.sh) so I'm really unsure what is the right thing to do.

We've already got some stuff in third_party/bazel; you could move the scripts there and nore in the README.google that you extracted them from xcrunwrapper (at a specific SHA, not `master`). Maybe fromberger would know a better thing to do.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-07 01:14:29_:

>>! In (earlier comment), zarko wrote:
>>>! In (earlier comment), salguarnieri wrote:
>>>>! In (earlier comment), zarko wrote:
>>> Is it possible to include a unit test that runs the extractor on a hello, worldish example?
>> 
>> I'll see if I can add one. I didn't see one for the cxx bazel extractor, should I create one for that also or did I miss it?
> 
> You're right; it doesn't look like we have that hooked up. (I'd initially thought that `cc_verifier_test` uses it, but reality differs.) It would be better to keep this diff ObjC-only and to consider adding one of the cxx extractor separately.
> 

Will do. I was hoping that you'd magically say there was something I could copy and paste, but no such luck :) I'll create a ticket for a cxx test as well so I/we don't forget.

>>> Do we need to put the Bazel-owned scripts in `third_party`?
>> Maybe. I carved apart a single script from them (https://github.com/bazelbuild/bazel/blob/master/src/tools/xcode/xcrunwrapper/xcrunwrapper.sh) so I'm really unsure what is the right thing to do.
> 
> We've already got some stuff in third_party/bazel; you could move the scripts there and nore in the README.google that you extracted them from xcrunwrapper (at a specific SHA, not `master`). Maybe fromberger would know a better thing to do.

Moved to third_party bazel and updated readme

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-07 01:15:03_:

  - move xcode helper scripts

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-08 22:28:20_:

switch to re2, put objc extractor helper functions into their own file to make testing easier, add unit tests, add an integration test for objc bazel extractor.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-08 22:31:56_:

*note that we won't run this in presubmit*

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-08 22:31:56_:

*note that we won't run this in presubmit*

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-08 22:31:56_:

>>! In (earlier comment), salguarnieri wrote:
>>>! In (earlier comment), zarko wrote:
>>>>! In (earlier comment), salguarnieri wrote:
>>>>>! In (earlier comment), zarko wrote:
>>>> Is it possible to include a unit test that runs the extractor on a hello, worldish example?
>>> 
>>> I'll see if I can add one. I didn't see one for the cxx bazel extractor, should I create one for that also or did I miss it?
>> 
>> You're right; it doesn't look like we have that hooked up. (I'd initially thought that `cc_verifier_test` uses it, but reality differs.) It would be better to keep this diff ObjC-only and to consider adding one of the cxx extractor separately.
>> 
> 
> Will do. I was hoping that you'd magically say there was something I could copy and paste, but no such luck :) I'll create a ticket for a cxx test as well so I/we don't forget.
> 
>>>> Do we need to put the Bazel-owned scripts in `third_party`?
>>> Maybe. I carved apart a single script from them (https://github.com/bazelbuild/bazel/blob/master/src/tools/xcode/xcrunwrapper/xcrunwrapper.sh) so I'm really unsure what is the right thing to do.
>> 
>> We've already got some stuff in third_party/bazel; you could move the scripts there and nore in the README.google that you extracted them from xcrunwrapper (at a specific SHA, not `master`). Maybe fromberger would know a better thing to do.
> 
> Moved to third_party bazel and updated readme


Added tests, should be ready for review now.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-09 22:16:22_:

nit: these global functions should start with a capital letter.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-09 22:16:22_:

nit: 2016 (here and elsewhere)

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-09 22:16:22_:

What does this function do when the script dies due to an error?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-09 22:16:22_:

Is there something in llvm::StringRef you could use for this?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-09 22:16:22_:

Also add the "manual" tag.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-09 22:43:34_:

  - address comments

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-09 22:44:07_:

Thankfully yes

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-09 22:44:07_:

The function should return the empty string in that case.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-12 22:44:29_:

  - Make script +x

**Comment by kythe** [comment]
_**zarko** wrote at 2016-09-12 23:26:29_:

Note #1385 (which this appears to tickle). Checking over the rest now.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-12 23:44:14_:

nit: you probably want ".objc.kindex" here and on line 39

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-12 23:44:14_:

There doesn't appear to be a dependency between this rule and objc_lib.{m,h}. Since I expect adding a direct dependency on objc_lib would be very inconvenient, consider adding objc_lib.{m,h} as data for this test.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-12 23:44:14_:

There are no assertions in either of the source files, so it's not clear that the entries ever hit the verifier. If I add a broken assertion (`//- bad`) to objc_lib.m, `:objc_lib_test` still succeeds.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-12 23:52:56_:

Aha, you're actually running the verifier on the .xa:
```
cat objc_lib_test
#!/bin/bash -e
set -o pipefail
gunzip -c kythe/cxx/extractor/testdata/objc_lib_test.compilation/kythe/cxx/extractor/testdata/objc_lib.xa.entries.gz | kythe/cxx/verifier/verifier --ignore_dups kythe/cxx/extractor/testdata/objc_lib.xa
```
This should probably be run on the `.m` and the `.h`. Maybe the `.xa` should be the `data` and the test source should be the `srcs`?

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-13 00:23:38_:

Added assertions

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-13 00:23:38_:

updated to do that. Let me know if I understood you correctly.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-13 00:24:06_:

  - Address comments
  - Update xa file for objective c.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-13 00:45:45_:

  - swap c++ for objc in strings

**Comment by kythe** [comment]
_**zarko** wrote at 2016-09-13 00:50:49_:

Thanks for hitting all the comments.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-09-13 00:50:49_

**Comment by kythe** [comment]
_**salguarnieri** closed via commit 455429c993582f05f59e46891b737cb82cf5497c at 2016-09-13 00:55:29_
